services/indexes/pvm/writer.go

```diff
@@ -410,67 +401,84 @@ func (w *Writer) indexTransaction(ctx services.ConsumerCtx, blkID ids.ID, tx txs
 			ChainID: w.chainID,
 		}
 		typ = models.TransactionTypeAddDelegator
-		err = w.InsertTransactionValidator(ctx, tx.ID(), castTx.Validator)
-		if err != nil {
-			return err
-		}
-		err = w.InsertTransactionBlock(ctx, tx.ID(), blkID)
+		err := w.InsertTransactionValidator(ctx, txID, castTx.Validator)
 		if err != nil {
 			return err
 		}
-		err = w.insertTransactionsRewardsOwners(ctx, tx.ID(), castTx.DelegationRewardsOwner, baseTx, castTx.StakeOuts)
+		err = w.insertTransactionsRewardsOwners(ctx, txID, castTx.DelegationRewardsOwner, baseTx, castTx.StakeOuts)
 		if err != nil {
 			return err
 		}
 	case *txs.CreateSubnetTx:
 		baseTx = castTx.BaseTx.BaseTx
 		typ = models.TransactionTypeCreateSubnet
-		err = w.InsertTransactionBlock(ctx, tx.ID(), blkID)
-		if err != nil {
-			return err
-		}
 	case *txs.CreateChainTx:
 		baseTx = castTx.BaseTx.BaseTx
 		typ = models.TransactionTypeCreateChain
-		err = w.InsertTransactionBlock(ctx, tx.ID(), blkID)
-		if err != nil {
-			return err
-		}
 	case *txs.ImportTx:
 		baseTx = castTx.BaseTx.BaseTx
 		ins = &avaxIndexer.AddInsContainer{
 			Ins:     castTx.ImportedInputs,
 			ChainID: castTx.SourceChain.String(),
 		}
 		typ = models.TransactionTypePVMImport
-		err = w.InsertTransactionBlock(ctx, tx.ID(), blkID)
-		if err != nil {
-			return err
-		}
 	case *txs.ExportTx:
 		baseTx = castTx.BaseTx.BaseTx
 		outs = &avaxIndexer.AddOutsContainer{
 			Outs:    castTx.ExportedOutputs,
 			ChainID: castTx.DestinationChain.String(),
 		}
 		typ = models.TransactionTypePVMExport
-		err = w.InsertTransactionBlock(ctx, tx.ID(), blkID)
-		if err != nil {
-			return err
-		}
 	case *txs.AdvanceTimeTx:
 		return nil
 	case *txs.RemoveSubnetValidatorTx:
-		return nil
+		baseTx = castTx.BaseTx.BaseTx
+		typ = models.TransactionTypeRemoveSubnetValidator
 	case *txs.TransformSubnetTx:
-		return nil
+		baseTx = castTx.BaseTx.BaseTx
+		typ = models.TransactionTypeTransformSubnet
 	case *txs.AddPermissionlessValidatorTx:
-		return nil
+		baseTx = castTx.BaseTx.BaseTx
+		outs = &avaxIndexer.AddOutsContainer{
+			Outs:    castTx.StakeOuts,
+			Stake:   true,
+			ChainID: w.chainID,
+		}
+		typ = models.TransactionTypeAddPermissionlessValidator
+		// TODO: Should this be reported for all subnetIDs?
+		if castTx.Subnet == constants.PrimaryNetworkID {
+			err := w.InsertTransactionValidator(ctx, txID, castTx.Validator)
+			if err != nil {
+				return err
+			}
+			// TODO: What to do about the different rewards owners?
+			err = w.insertTransactionsRewardsOwners(ctx, txID, castTx.ValidatorRewardsOwner, baseTx, castTx.StakeOuts)
+			if err != nil {
+				return err
+			}
+		}
 	case *txs.AddPermissionlessDelegatorTx:
-		return nil
+		baseTx = castTx.BaseTx.BaseTx
+		outs = &avaxIndexer.AddOutsContainer{
```

Comment:
I think this should only be defined if castTx.Subnet == constants.PrimaryNetworkID